### PR TITLE
adding flag option to only pad the last dim

### PR DESCRIPTION
**Pull request by btolooshams**
Added a pad_only_last_dim in DomainPadding class (default False). Updated FNO class to have the flag domain_padding_only_last_dim (default False). If set True, then padding is only applied to the last dim of the input.

### CODE REVIEW

**Comment by JeanKossaifi** [comment]
Thanks @btolooshams. Let's make it more general, how about having two use cases:
- padding is True, in that case set padding = range(...) such that we pad all dimensions
- Padding is a list of dimensions (for your use case would be [4], in that case we just pad those dimensions.

The only question is whether we account batch-size and channels in the dimensions (dim 0 is then batch size). Otherwise 0 would be the first spatial dim. 
  

What do you think? Could you make the change for the padding and update the docstrings in the models?

**Comment by btolooshams** [comment]
I have modified the previous pool request. The "pad_only_last_dim" is now changed to "padding_dim". If True, all dimensions (0,1,2) excluding batch and channel with be padded, if "padding_dim" is a list, then those dimensions in the list will only be padded.

An additional fix --> the padding in F is applied in reverse order, so I have added a line to reverse the padding. Padding was being wrongly applied before.

**Comment by btolooshams** [comment]
The current implementation excludes batch and ch dim (0,1) in counting with dimension to pad. 0 is the first spatial dimension.

**Comment by JeanKossaifi** [comment]
Thank you @btolooshams, this is great! Could we call it just `padding`? 
Would be great to add a quick test case in the unit-tests for sanity and to make sure it remains correct in the future?

File: neuralop/layers/tests/test_padding.py [hunk 1]
```diff
@@ -3,16 +3,44 @@
...
+
     data = torch.randn((2, 3, 10, 10))
-    padder = DomainPadding(0.2, mode)
+    padder = DomainPadding(padding_dim, mode)
     padded = padder.pad(data)
 
     target_shape = list(padded.shape)
-    target_shape[-1] = target_shape[-2] = out_size[mode]
+    for ctr in range(1,3):
+        target_shape[-ctr] = out_size[mode][-ctr]
```
Comment (JeanKossaifi) [inline neuralop/layers/tests/test_padding.py:19]: Can we just write this out to make it easier to read with a small inline comment? ALso I don't know what `ctr` means, good to use clear names

File: neuralop/layers/tests/test_padding.py [hunk 2]
```diff
@@ -3,16 +3,44 @@
...
+@pytest.mark.parametrize('padding_dim', [0.2, [0.1, 0.2]])
+def test_DomainPadding_2d(mode, padding_dim):
+    if isinstance(padding_dim, float):
+        out_size =  {'one-sided': [12, 12], 'symmetric': [14, 14]}
+    else:
+        out_size =  {'one-sided': [11, 12], 'symmetric': [12, 14]} 
+
     data = torch.randn((2, 3, 10, 10))
-    padder = DomainPadding(0.2, mode)
+    padder = DomainPadding(padding_dim, mode)
```
Comment (JeanKossaifi) [inline neuralop/layers/tests/test_padding.py:14]: Why do we call it padding_dim here?

File: neuralop/layers/padding.py [hunk 3]
```diff
@@ -52,33 +57,50 @@ def pad(self, x):
...
             # padded by the amount mention at the first position of the 'padding' vector.
             # The details about F.pad can be found here : https://pytorch.org/docs/stable/generated/torch.nn.functional.pad.html
 
             if self.padding_mode == 'symmetric':
                 # Pad both sides
-                unpad_indices = (Ellipsis, ) + tuple([slice(p, -p, None) for p in output_pad[::-1] ])
+                unpad_list = list()
+                for p in output_pad[::-1]:
+                    if p == 0:
+                        unpad_amount_pos = None
```
Comment (JeanKossaifi) [inline neuralop/layers/padding.py:79]: Why not leave 0 in this case?

File: neuralop/layers/padding.py [hunk 4]
```diff
@@ -52,33 +57,50 @@ def pad(self, x):
...
+                        unpad_amount_neg = None
+                    else:
+                        unpad_amount_pos = p
+                        unpad_amount_neg = -p
+                    unpad_list.append(slice(unpad_amount_pos, unpad_amount_neg, None))
```
Comment (JeanKossaifi) [inline neuralop/layers/padding.py:84]: I actually don't mind spelling it out but can we call it padding_start and padding_end or something clearer? :)

File: neuralop/layers/padding.py [hunk 5]
```diff
@@ -39,8 +40,12 @@ def pad(self, x):
         """
         resolution = x.shape[2:]
 
+        # if domain_padding is list, then to pass on
         if isinstance(self.domain_padding, (float, int)):
             self.domain_padding = [float(self.domain_padding)]*len(resolution)
+
+        assert len(self.domain_padding) == len(resolution)
```
Comment (JeanKossaifi) [inline neuralop/layers/padding.py:47]: Good check to have! Can you add a message if that's not the case to explain the issue to users?

File: neuralop/layers/tests/test_padding.py [hunk 6]
```diff
@@ -3,16 +3,44 @@
...
+        out_size =  {'one-sided': [12, 12, 12], 'symmetric': [14, 14, 14]}
+    else:
+        out_size =  {'one-sided': [11, 10, 12], 'symmetric': [12, 10, 14]} 
+
+    data = torch.randn((2, 3, 10, 10, 10))
+    padder = DomainPadding(padding_dim, mode)
+    padded = padder.pad(data)
+
+    target_shape = list(padded.shape)
+    for ctr in range(1,4):
```
Comment (JeanKossaifi) [inline neuralop/layers/tests/test_padding.py:39]: Same comment as above - ideally tests should be self explanatory and easy to get by just reading the code!

**Review by JeanKossaifi — COMMENTED** [review]
Thanks @btolooshams! I like the API like this, I left a few comments, mostly just style.

File: neuralop/layers/tests/test_padding.py (hunk 1 above)
Comment (btolooshams) [reply neuralop/layers/tests/test_padding.py:19]: "ctr" is changed to "pad_dim", and comment added.

File: neuralop/layers/tests/test_padding.py (hunk 2 above)
Comment (btolooshams) [reply neuralop/layers/tests/test_padding.py:14]: padding_dim is renamed to "padding"

File: neuralop/layers/padding.py (hunk 3 above)
Comment (btolooshams) [reply neuralop/layers/padding.py:79]: slice gives an error for p=0. So if padding is 0, p is set to None, hence no slicing is done to unpad list.

File: neuralop/layers/padding.py (hunk 4 above)
Comment (btolooshams) [reply neuralop/layers/padding.py:84]: name modified :)

File: neuralop/layers/padding.py (hunk 5 above)
Comment (btolooshams) [reply neuralop/layers/padding.py:47]: Explanation is added. "domain_padding length must match the number of spatial/time dimensions (excluding batch, ch)"

**Comment by JeanKossaifi** [comment]
Thank you @btolooshams, this is great, merging!
